Tablet keyboard

bhanu-dev82 opened the pull request:
<!---
Thank you for your pull request! 🚀
-->

### Contributor checklist

<!-- Please replace the empty checkboxes [] below with checked ones [x] accordingly. -->

-   [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch
-   [x] I have tested my code with the `./gradlew lintKotlin detekt test` command as directed in the [testing section of the contributing guide](https://github.com/scribe-org/Scribe-Android/blob/main/CONTRIBUTING.md#testing)

---

### Description

This pull request addresses issue #280, implementing layout improvements for different screen sizes and adding enhanced keyboard functionality.

**Key Changes:**

* **Resource Updates for Different Screen Sizes:**
    * `app/src/main/res/values-land/dimens.xml`: Added values for landscape mode.
    * `app/src/main/res/values-w600dp/dimens.xml`: Added values for medium tablets.
    * `app/src/main/res/values-w1240dp/dimens.xml`: Added values for larger tablets.

* **Enhanced Keyboard Functionality:**
    * `app/src/main/java/be/scri/services/EnglishKeyboardIME.kt`: Added custom functions for Caps Lock, Shift, Tab, and Left/Right key handling.
    * `app/src/main/java/be/scri/views/KeyboardView.kt`: Updated label handling for Caps Lock, Shift, and Tab keys.
    * `app/src/main/java/be/scri/helpers/KeyboardBase.kt`: Defined custom keys (Caps Lock, Tab, Left/Right arrows).


**Testing:**

The changes have been tested using the following command:

```bash
./gradlew lintKotlin detekt test
```

**Screenshot:**

Tablet landscape mode 
![image](https://github.com/user-attachments/assets/0db69a7f-2d95-4cec-8c22-ca411aa74761)

Tablet Portrait mode 
![image](https://github.com/user-attachments/assets/4f68e098-b15c-4219-8274-3d73abd27989)

angrezichatterbox commented:
Thanks for the PR @bhanu-dev82. 
Could I get to know which device you are testing this on. Currently on the Pixel Tablet API 35 it doesn't show up the extra keys. Could I get to know from which device onwards it would be using the keyboard you have depicted in the PR description.

bhanu-dev82 commented:
> Thanks for the PR @bhanu-dev82. Could I get to know which device you are testing this on. Currently on the Pixel Tablet API 35 it doesn't show up the extra keys. Could I get to know from which device onwards it would be using the keyboard you have depicted in the PR description.

Hello @angrezichatterbox i have also used pixel tablet api 35, the changes are only for scrib english. i will soon aad the video instructions also. Thankyou

bhanu-dev82 commented:
> > > Thanks for the PR @bhanu-dev82. Could I get to know which device you are testing this on. Currently on the Pixel Tablet API 35 it doesn't show up the extra keys. Could I get to know from which device onwards it would be using the keyboard you have depicted in the PR description.
> > 
> > 
> > Hello @angrezichatterbox i have also used pixel tablet api 35, the changes are only for scribe english. i will soon aad the video instructions also. Thankyou
> 
> Video instruction aren't required. Thanks. We will get to review it soon :)

is the new layout visible i have tested it again by uninstalling and installing it. and it is showing up properly, 
![image](https://github.com/user-attachments/assets/fdae3005-9e20-4962-9b3e-da7d44d5ad8e)

bhanu-dev82 commented:
> Can the key sizes be increased a bit for the landscape keyboard The height of the keys to be specific.
> 
> Can the key handler be moved to the helper directory and used within the EnglishKeyboardIME as well. Thanks for making the key handlers to a separate class it is really helpful moving forward.

i will implement both of this today only, how should the new pull request be made? should i delete this one and create a new pull request?

angrezichatterbox commented:
You can push your changes to the same branch. You need not close this PR.

bhanu-dev82 commented:
> You can push your changes to the same branch. You need not close this PR.

To make the inner class a separate file i need to update GeneralKeyboardIME.kt  to have some variables and function to use internal instead of protected, also i need to have this file inside the package : be.scri.services as KeyHandler.kt.    is fine if i proceed with this changes ? this approach will keep the encapsulation intact.

angrezichatterbox commented:
> > You can push your changes to the same branch. You need not close this PR.
> 
> To make the inner class a separate file i need to update GeneralKeyboardIME.kt to have some variables and function to use internal instead of protected, also i need to have this file inside the package : be.scri.services as KeyHandler.kt. is it fine if i proceed with this changes ? this approach will keep the encapsulation intact.

Yes it is fine. I would look into the changes once you have made it. The KeyboardIME were getting larger and it would be better if we have it made smaller.

bhanu-dev82 commented:
> > > You can push your changes to the same branch. You need not close this PR.
> > 
> > 
> > To make the inner class a separate file i need to update GeneralKeyboardIME.kt to have some variables and function to use internal instead of protected, also i need to have this file inside the package : be.scri.services as KeyHandler.kt. is it fine if i proceed with this changes ? this approach will keep the encapsulation intact.
> 
> Yes it is fine. I would look into the changes once you have made it. The KeyboardIME were getting larger and it would be better if we have it made smaller.

The file KeyHandler has been successfully made  and also tested the changes to ensure everything works,  for the key height what i have done is calculated the ration in the current implementation and then implemented the same in landscape mode there was a 20 percent increase in key height as seen below in the attached image : 
![image](https://github.com/user-attachments/assets/326ffe55-0120-4fb1-bee8-67974a3b02fc)

angrezichatterbox commented:
It would be better to move the handler to helpers folder right. This would help keep the structure and contents of folders intact. Other than that the changes look good to me.

andrewtavis commented:
Can we make the Scribe key a bit wider in this layout? Ideally the width of the tab key?

bhanu-dev82 commented:
> It would be better to move the handler to helpers folder right. This would help keep the structure and contents of folders intact. Other than that the changes look good to me.

Relocating the handler would require making protected variables public, which would break encapsulation. For this reason, I have kept it within the same package, i.e., services. If modifying the variable's access level is not a concern, I can proceed with moving the file to helpers.

I would appreciate some guidance on the best approach in this scenario. Let me know your thoughts!

bhanu-dev82 commented:
> Can we make the Scribe key a bit wider in this layout? Ideally the width of the tab key?

i am finding solution for it. we need to have changes for tablet / landscape mode only right?

andrewtavis commented:
Yes I'd say for now table / landscape is the important one to change, and we can get to the others later :)

Thanks for looking into this, @bhanu-dev82! 😊

angrezichatterbox commented:
> > It would be better to move the handler to helpers folder right. This would help keep the structure and contents of folders intact. Other than that the changes look good to me.
> 
> Relocating the handler would require making protected variables public, which would break encapsulation. For this reason, I have kept it within the same package, i.e., services. If modifying the variable's access level is not a concern, I can proceed with moving the file to helpers.
> 
> I would appreciate some guidance on the best approach in this scenario. Let me know your thoughts!

I would expect the Key handler class to be taking in the private IME of type GeneralKeyboardIME abstract class. Could you tell which variables would be the ones affecting the encapsulation in that case.

bhanu-dev82 commented:
> > > It would be better to move the handler to helpers folder right. This would help keep the structure and contents of folders intact. Other than that the changes look good to me.
> > 
> > 
> > Relocating the handler would require making protected variables public, which would break encapsulation. For this reason, I have kept it within the same package, i.e., services. If modifying the variable's access level is not a concern, I can proceed with moving the file to helpers.
> > I would appreciate some guidance on the best approach in this scenario. Let me know your thoughts!
> 
> I would expect the Key handler class to be taking in the private IME of type GeneralKeyboardIME abstract class. Could you tell which variables would be the ones affecting the encapsulation in that case.

Pardon me i didn't get what you are saying, can you guide me on what should be done to handle this keyhandler.kt?

angrezichatterbox commented:
Sorry if I wasn't clear enough. So currently the Keyhandler.kt class present in it takes only ime of type EnglishKeyboardIME. This could be replaced with the abstract class it is based on that is GeneralKeyboardIME right so that it could accept all the different IME. 
Can I know how would encapsulation be broken if the class is moved to another package. The functions would remain private right.

bhanu-dev82 commented:
> Yes I'd say for now table / landscape is the important one to change, and we can get to the others later :)
> 
> Thanks for looking into this, @bhanu-dev82! 😊

I have made the scribe button modular with the help of dimens.xml so now it can be increased or decreased in size as we require. i have tried different values till it is of similar width to that of tab key, here is an attached image to show the new UI.
![Screenshot_20250228_185834](https://github.com/user-attachments/assets/a0f6790e-e441-4f9d-8ef8-7be7ef66e75c)

andrewtavis commented:
Looking wonderful, @bhanu-dev82! Thanks for the quick corrections :)

bhanu-dev82 commented:
> Sorry if I wasn't clear enough. So currently the Keyhandler.kt class present in it takes only ime of type EnglishKeyboardIME. This could be replaced with the abstract class it is based on that is GeneralKeyboardIME right so that it could accept all the different IME. Can I know how would encapsulation be broken if the class is moved to another package. The functions would remain private right.

pardon me again there was some misinterpretation from my side, i have moved the KeyHandler.kt to helpers. and about generalization of the file, i will soon find a solution for it, i am thinking of a modular approach to improve the Code Reusability

bhanu-dev82 commented:
> > Sorry if I wasn't clear enough. So currently the Keyhandler.kt class present in it takes only ime of type EnglishKeyboardIME. This could be replaced with the abstract class it is based on that is GeneralKeyboardIME right so that it could accept all the different IME. Can I know how would encapsulation be broken if the class is moved to another package. The functions would remain private right.
> 
> pardon me again there was some misinterpretation from my side, i have moved the KeyHandler.kt to helpers. and about generalization of the file, i will soon find a solution for it, i am thinking of a modular approach to improve the Code Reusability.

bhanu-dev82 commented:
> Sorry if I wasn't clear enough. So currently the Keyhandler.kt class present in it takes only ime of type EnglishKeyboardIME. This could be replaced with the abstract class it is based on that is GeneralKeyboardIME right so that it could accept all the different IME. Can I know how would encapsulation be broken if the class is moved to another package. The functions would remain private right.

I have moved the Keyhandler.kt file to the helpers package and generalized it to support all IMEs. Additionally, I have added icons for all the new keys. I believe this covers everything for this pull request. Please let me know if anything is missing. For each IME, we can create separate issues to discuss and implement necessary changes, as the overall structure is now in place—we would just need to define the key mappings.
![image](https://github.com/user-attachments/assets/c50be00a-274a-4a32-92d9-4d3071352fee)
With caps lock on
![image](https://github.com/user-attachments/assets/d0ae8c84-f648-4b2f-8c75-48192b6f3434)

angrezichatterbox commented:
This could be merged in right. For the other languages we could have a checklist in the issue for the language completed and to be worked on right. 

We could also have an issue to improve the styling of command bar for tablets once this is merged in.

andrewtavis commented:
I agree that we can have a checklist in the issue description of #280 rather than making new issues for each keyboard. Do you want to edit the description and add that in, @angrezichatterbox? 

Aside from that, happy to do a review myself sometime today 😊

angrezichatterbox commented:
> I agree that we can have a checklist in the issue description of #280 rather than making new issues for each keyboard. Do you want to edit the description and add that in, @angrezichatterbox?

I will add that in the issue description

bhanu-dev82 commented:
> This could be merged in right. For the other languages we could have a checklist in the issue for the language completed and to be worked on right.
> 
> We could also have an issue to improve the styling of command bar for tablets once this is merged in.

I have reviewed each language layout and determined that updating the keyboard with a tablet-specific layout for each language requires only one XML file per language. since most layouts share common elements, I can complete all eight XML files today.
Additionally, I will include eight screenshots of the updated layouts to ensure this issue is fully addressed.

andrewtavis commented:
Great working with you on this, @bhanu-dev82! 😊

bhanu-dev82 commented:
> Great working with you on this, @bhanu-dev82! 😊

German keyboard is what is left i am testing it currently will push in a minute

bhanu-dev82 commented:
I have updated all the languages for tablet layout. Kindly provide feedback. The attached Images are the updated layouts.

English :
![image](https://github.com/user-attachments/assets/df774925-a845-4405-8c71-e7d9abaff753)

French : 
![image](https://github.com/user-attachments/assets/72cd6ac1-d12c-4e1e-b003-e6c6e2a4d7cf)

German : 
![image](https://github.com/user-attachments/assets/2eee1a53-5c4e-4767-a675-58b587d2229d)

Italian : 
![image](https://github.com/user-attachments/assets/d92fab1d-8329-4e06-aefd-9380652c5c46)

Portuguese :  
![image](https://github.com/user-attachments/assets/adcd5725-cb2b-4d3d-99da-084693ff94b9)

Russian : 
![image](https://github.com/user-attachments/assets/37e7f668-73ae-4ab4-8c32-a956538ba7a1)

Spanish : 
![image](https://github.com/user-attachments/assets/6563bc34-f5ed-4d0c-8031-02b3d8e19145)

Swedish : 
![image](https://github.com/user-attachments/assets/dde2b9a3-3e12-4e59-9641-517ca4d22b1b)

andrewtavis reviewed (approved):
Commits on my end were to switch the background of the emoji suggestions to transparent :) Thanks for the amazing work here, @bhanu-dev82, and also for the detailed review, @angrezichatterbox! Appreciate the collaboration here from both of you 😊 We can close this and any further changes can be in individual mini issues closer to release 🚀
